### DISCUSSION TITLE
Unable to make custom input component work.

### DISCUSSION THREAD

**Discussion by Bharateshwar** [react-hook-form/react-hook-form#2531] [Q&A] [answered]
Hi ... I have created a sort of custom component of my own and then I'm putting it inside the form. 
But I'm not sure of how to actually make this work. 

My custom component is :

```
`

import React, { useState } from 'react'
import { InputAdornment } from '@material-ui/core'
import { DatePicker } from "@material-ui/pickers";
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import TodayIcon from '@material-ui/icons/Today';
import DateFnsUtils from '@date-io/date-fns';
import './DateInput.css'



const DateInput = (props) => {

    const [selectedDate, handleDateChange] = useState(new Date());

    return (

        
            <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <DatePicker className={props.className}
                    value={selectedDate} onChange={handleDateChange} variant="dialog"
                    inputVariant="outlined"
                    InputProps={{
                        startAdornment: (
                            <InputAdornment position="start">
                                <TodayIcon />
                            </InputAdornment>
                        )
                    }}
                    disablePast
                    format="dd MMMM yyyy"
                />
            </MuiPickersUtilsProvider>
    )

}


export default DateInput

`

```

and I'm using it Like

```
<form>
<Controller
                                as={<DateInput className="date-input" />}
                                control={control}
                                name="date"
                                defaultValue=""
                            />
<form>

```

I tried using Controller, tried attaching control via props but I am unable to get it to work.

Help Please!

**Comment by bluebill1049** [comment, marked as answer]
you should have read at the Controller and have a good understanding of it first: https://react-hook-form.com/api#Controller
<img width="964" alt="Screen Shot 2020-08-09 at 9 19 42 pm" src="https://user-images.githubusercontent.com/10513364/89730876-0d30db80-da86-11ea-9b22-bddf1684ceca.png">

Controller pass down props, and those props should plug them in with your input component, in your example `DatePicker` should receive all the props from the Controller: value, onChange and onBlur.

**Reply by Bharateshwar** [reply]
` you can pass it to the Controller component directly and it will take care of forwarding the prop for you.`  <-- this line had me a bit confused 

I understand that the props are passed by the controller ... but the props are passed to the top level component of the custom component. 
I thought it handles passing the props to underlying core input element directly irrespective of other parent components around it. I had to spread the props to my `DatePicker` manually as it is wrapped in  `MuiPickersUtilsProvider`. 

Maybe you could include it in the documents. I think you should ... but you know better. I'm just a rookie.  :)


The mistake I was doing was handling change in my custom component itself, your reply helped.



Thanks BTW. 

I appreciate your quick responses. One of the reasons I always use react-hook-form over other libraries. 

Thanks again. :)

**Reply by bluebill1049** [reply]
hey @Bharateshwar much-apprecated ur support and feedback. (you are definitely not a rookie ❤️) In fact, feel free to contribute to the doc, having a better description will not just help you but others. 

> you can pass it to the Controller component directly and it will take care of forwarding the prop for you.

or feel free to let me know how to improve on above as well.

**Reply by Bharateshwar** [reply]
I was just saying that in the docs you have mentioned that the Controller takes care of forwarding the props, but this is only helpful and true in cases when the parent element of the Component we intent to use with Controller is supposed to have those props (onChange, etc). 

Automatic forwarding does not help when the parent element of our custom component (as in my example) is not supposed to have any props or methods like onChange. 

For Example: 

```
                              <Controller
                                as={<DateInput className="date-input" />}
                                control={control}
                                name="date"
                                defaultValue={selectedDate}
                                value={selectedDate}
                            /> 
```

Just doing this is not enough as my custom Component DateInput has a DatePicker Component in it which is wrapped in another component. And the props forwarded by Controller are supposed to be passed onto the DatePicker and not to the parent component wrapping the DatePicker. 

I had to use {...props} to spread the props manually to the DatePicker. 

I hope this makes sense and I was able to explain what I meant. :)
